pkg/bmc/bmc.go

```diff
@@ -347,3 +367,177 @@ func redfishGetSystemSecureBoot(redfishClient *gofish.APIClient, systemIndex int
+		glog.V(100).Infof("Failed to create a new SSH session: %v", err)
+
+		return nil, fmt.Errorf("failed to create a new ssh session: %w", err)
+	}
+
+	return session, nil
+}
+
+// RunCLICommand runs a CLI command in the BMC's console. This method will block until the command
+// has finished, and its output is copied to stdout and/or stderr if applicable. If combineOutput is true,
+// stderr content is merged in stdout. The timeout param is used to avoid the caller to be stuck forever
+// in case something goes wrong or the command is stuck.
+func (bmc *BMC) RunCLICommand(cmd string, combineOutput bool, timeout time.Duration) (
+	stdout string, stderr string, err error) {
+	glog.V(100).Infof("Running CLI command in BMC's CLI: %s", cmd)
+
+	sshSession, err := bmc.CreateCLISSHSession()
+	if err != nil {
+		glog.V(100).Infof("Failed to connect to CLI: %v", err)
+
+		return "", "", fmt.Errorf("failed to connect to CLI: %w", err)
+	}
+
+	defer sshSession.Close()
+
+	var stdoutBuffer, stderrBuffer bytes.Buffer
+	if !combineOutput {
+		sshSession.Stdout = &stdoutBuffer
+		sshSession.Stderr = &stderrBuffer
+	}
+
+	var combinedOutput []byte
+
+	errCh := make(chan error)
+	go func() {
+		var err error
+		if combineOutput {
+			combinedOutput, err = sshSession.CombinedOutput(cmd)
+		} else {
+			err = sshSession.Run(cmd)
+		}
+		errCh <- err
+	}()
+
+	timeoutCh := time.After(timeout)
+
+	select {
+	case <-timeoutCh:
+		glog.V(100).Info("CLI command timeout")
+
+		return stdoutBuffer.String(), stderrBuffer.String(), fmt.Errorf("timeout running command")
+	case err := <-errCh:
+		glog.V(100).Info("Command run error: %v", err)
+
+		if err != nil {
+			return stdoutBuffer.String(), stderrBuffer.String(), fmt.Errorf("command run error: %w", err)
+		}
+	}
+
+	if combineOutput {
+		return string(combinedOutput), "", nil
+	}
+
+	return stdoutBuffer.String(), stderrBuffer.String(), nil
+}
+
+// OpenSerialConsole opens the serial console port. The console is tunneled in an underlying (CLI)
+// ssh session that is opened in the BMC's ssh server. If openConsoleCliCmd is
+// provided, it will be sent to the BMC's cli. Otherwise, a best effort will
+// be made to run the appropriate cli command based on the system manufacturer.
+func (bmc *BMC) OpenSerialConsole(openConsoleCliCmd string) (io.Reader, io.WriteCloser, error) {
+	if bmc.sshSessionForSerialConsole != nil {
+		return nil, nil, fmt.Errorf("there is already a serial console opened for this BMC")
+	}
+
+	cliCmd := openConsoleCliCmd
+	if cliCmd == "" {
+		// no cli command to get console port was provided, try to guess based on
+		// manufacturer.
+		manufacturer, err := bmc.SystemManufacturer()
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get redfish system manufacturer: %w", err)
+		}
+
+		var found bool
+		if cliCmd, found = cliCmdSerialConsole[manufacturer]; !found {
+			return nil, nil, fmt.Errorf("cli command to get serial console not found for manufacturer %v", manufacturer)
```

Comment:
Please add glog

Reply:
Done in https://github.com/openshift-kni/eco-goinfra/pull/419/commits/74bdadb9f818f05f583fe09fd8569c0f403eb56e